commons/src/test/java/io/aiven/kafka/tieredstorage/commons/SegmentManifestProviderTest.java

```diff
@@ -0,0 +1,119 @@
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class SegmentManifestProviderTest {
+    static final ObjectKey OBJECT_KEY = new ObjectKey("");
+
+    static final ObjectMapper MAPPER = new ObjectMapper();
+
+    static {
+        MAPPER.registerModule(new Jdk8Module());
+    }
+
+    static final String MANIFEST =
+        "{\"version\":\"1\","
+            + "\"chunkIndex\":{\"type\":\"fixed\",\"originalChunkSize\":100,"
+            + "\"originalFileSize\":1000,\"transformedChunkSize\":110,\"finalTransformedChunkSize\":110},"
+            + "\"compression\":false}";
+
+    static final Uuid TOPIC_ID = Uuid.METADATA_TOPIC_ID;  // string representation: AAAAAAAAAAAAAAAAAAAAAQ
+    static final Uuid SEGMENT_ID = Uuid.ZERO_UUID;  // string representation: AAAAAAAAAAAAAAAAAAAAAA
+    static final RemoteLogSegmentMetadata REMOTE_LOG_METADATA = new RemoteLogSegmentMetadata(
+        new RemoteLogSegmentId(
+            new TopicIdPartition(TOPIC_ID, new TopicPartition("topic", 7)),
+            SEGMENT_ID
+        ),
+        23L, 2000L, 0, 0, 0, 1, Map.of(0, 0L));
+
+    @Mock
+    FileFetcher fetcher;
+
+    SegmentManifestProvider provider;
+
+    @BeforeEach
+    void setup() {
+        provider = new SegmentManifestProvider(OBJECT_KEY, 1000L, null, fetcher, MAPPER);
+    }
+
+    @Test
+    void unboundedShouldBeCreated() {
+        assertThatNoException()
+            .isThrownBy(() -> new SegmentManifestProvider(OBJECT_KEY, null, Duration.ofMillis(1), fetcher, MAPPER));
+    }
+
+    @Test
+    void withoutRetentionLimitsShouldBeCreated() {
+        assertThatNoException()
+            .isThrownBy(() -> new SegmentManifestProvider(OBJECT_KEY, 1L, null, fetcher, MAPPER));
+    }
+
+    @Test
+    void shouldReturnAndCache() throws StorageBackEndException, IOException {
+        final String key = "topic-AAAAAAAAAAAAAAAAAAAAAQ/7/00000000000000000023-AAAAAAAAAAAAAAAAAAAAAA.rsm-manifest";
+        when(fetcher.fetch(key))
+            .thenReturn(new ByteArrayInputStream(MANIFEST.getBytes()));
+        final SegmentManifestV1 expectedManifest = new SegmentManifestV1(
+            new FixedSizeChunkIndex(100, 1000, 110, 110),
+            false, null
+        );
+        assertThat(provider.get(REMOTE_LOG_METADATA)).isEqualTo(expectedManifest);
+        assertThat(provider.get(REMOTE_LOG_METADATA)).isEqualTo(expectedManifest);
+        verify(fetcher).fetch(key);
+    }
+
+    @Test
+    void shouldPropagateStorageBackEndException() throws StorageBackEndException {
+        when(fetcher.fetch(anyString()))
+            .thenThrow(new StorageBackEndException("test"));
+        assertThatThrownBy(() -> provider.get(REMOTE_LOG_METADATA))
+            .isInstanceOf(StorageBackEndException.class)
+            .hasMessage("test");
```

Comment:
Same for IOException/JSON related exceptions?

Reply:
Yes, done this